components/centraldashboard/public/components/utilities-mixin_test.js

```diff
@@ -1,37 +1,117 @@
-        expect(testObject.equals('abc', 'abc', 'abc')).toBe(true);
-        expect(testObject.equals('abc', 'abc', 'ABC')).toBe(false);
+        const {equals} = testObject;
+        expect(equals()).toBe(true);
+        expect(equals(1, '1')).toBe(false);
+        expect(equals(1, 1, 1)).toBe(true);
+        expect(equals('abc', 'abc', 'abc')).toBe(true);
+        expect(equals('abc', 'abc', 'ABC')).toBe(false);
+    });
+
+    it('Checks empty-ness on objects correctly', () => {
+        const {empty} = testObject;
+        expect(empty()).toBe(true);
+        expect(empty(0)).toBe(true);
+        expect(empty({})).toBe(true);
+        expect(empty([])).toBe(true);
+        expect(empty(new Set([]))).toBe(true);
+
+        expect(empty(1)).toBe(false);
+        expect(empty({a: 1})).toBe(false);
+        expect(empty([1])).toBe(false);
+        expect(empty(new Set([1]))).toBe(false);
+        expect(empty(new Error())).toBe(false);
+        expect(empty(new CustomEvent('test'))).toBe(false);
     });
 
     it('Builds links with valid query parameters', () => {
-        expect(testObject.buildHref('/foo', null)).toBe('/foo');
-        expect(testObject.buildHref('/foo?other_param=value', {}))
+        const {buildHref} = testObject;
+        expect(buildHref('/foo', null)).toBe('/foo');
+        expect(buildHref('/foo?other_param=value', {}))
             .toBe('/foo?other_param=value');
-        expect(testObject.buildHref('/foo?other_param=value', {ns: 'bar'}))
+        expect(buildHref('/foo?other_param=value', {ns: 'bar'}))
             .toBe('/foo?other_param=value&ns=bar');
     });
+
+    it('FireEvent correctly fires Event-Obj or custom events', () => {
+        const fireEvent = testObject.fireEvent.bind(testObject);
+        let lastEvent;
+        spyOn(testObject, 'dispatchEvent').and.callFake((ev) => {
+            lastEvent = {name: ev.type, detail: ev.detail, raw: ev};
+        });
+        fireEvent('foo');
+        expect(lastEvent.name).toBe('foo');
+        expect(lastEvent.detail).toEqual(null);
+
+        fireEvent('bar', {ap: 1});
+        expect(lastEvent.name).toBe('bar');
+        expect(lastEvent.detail).toEqual({ap: 1});
+
+        const ev = new Event('click');
+        fireEvent(ev);
+        expect(lastEvent.raw).toBe(ev);
+    });
+
+    it('Close Toast can correctly scan ancestry to the toast element', () => {
+        const {closeToast} = testObject;
+        const dom = html`
+            <paper-toast>
+                <span id='s1'>
+                    <span id='s2'>
+                        <span id='s3'></span>
+                    </span>
+                </span>
+            </paper-toast>
+        `;
+        const toast = dom.content.children[0];
+        const [s1, s2, s3] = [1, 2, 3]
+            .map((i) => toast.querySelector(`#s${i}`));
+        [toast, s1, s2, s3].forEach((el) => {
+            el.onclick = closeToast;
+        });
+        toast.close = () => {
```

Comment:
I'm not intimately familiar with the `paper-toast` object but it feels a bit unusual that you are overriding and adding new methods to its implementation. Your tests are then validating your own overridden logic and not the interface you are testing.

Is there a reason you couldn't rely on the `opened` property?

https://www.webcomponents.org/element/@polymer/paper-toast/elements/paper-toast

Reply:
That's because this test uses a non-dom base class w/o a test-fixture. So we actually never modify or touch the dom with any of these tests. Thus I never even load paper-toast (though this may work in this test). I just need to make sure that the test does call the `close` method on the correct parent in the `document-fragment` (or fake dom)

Reply:
To add to this, the goal is **not to test `paper-toast`** but rather to verify that a `.close()` call was made on said element.

Reply:
OK this makes sense now. I had just assumed this was using the text fixture structure..

You can then simplify it to:
```
toast.close = jasmine.createSpy();
toast.click();
s1.click();
s2.click();
s3.click();
expect(toast.close).toHaveBeenCalledTimes(3);
```

Reply:
I initially planned on doing it that way, but this test is slightly ambiguous since it does not ascertain which clicks actually caused the `.close()` calls. Do you think that's fair to ignore in favor of brevity?